### PR TITLE
Add support for ARM64 Docker images for all Beats

### PR DESCRIPTION
**Pull request by kvch**
This PR adds support for building ARM64 Docker images for all Beats on ARM64 machines.

To limit what packages should be built I have added the new environment variable named `PACKAGES`. It expects a comma-separated list of packages types. Available options: rpm, deb, tgz, dmg, docker, zip. If nothing is specified all available package types are built for the given platform.

This PR relies on the newly added crossbuild image with the suffix `base-arm-debian9`. This image is automatically selected if the runtime of the builder machine is ARM64.

A possible enhancement to this solution is described in #23775.

### CODE REVIEW

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:164]:
```suggestion
                    return params.arm
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:175]:
```suggestion
                  withGithubNotify(context: "Packaging linux/arm64 ${BEATS_FOLDER}") {
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:180]: It's not yet supported:

```suggestion
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:169]:
```suggestion
                  PACKAGES = "docker"
                  GOARCH = "arm64"
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:3]:
```suggestion
@Library('apm@current') _
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:24]:
```suggestion
    PIPELINE_LOG_LEVEL = "INFO"
```

**Comment by kvch** [comment]
Requires https://github.com/elastic/golang-crossbuild/pull/75

Otherwise packaging on ARM should be working after the builder selection is finalized.

**Comment by jsoriano** [comment]
@kvch I wonder if we can use `FROM --platfrom=...` to don't depend on `buildx` or additional images.

For example this Dockerfile generates a `linux/arm64` image with mage for arm64 from an amd64 machine, without buildx or special images:
```
FROM golang:1.15 AS build
ENV GOARCH=arm64 GOOS=linux
RUN go get -u -d github.com/magefile/mage
RUN go install github.com/magefile/mage

FROM --platform=linux/arm64 debian:stable-slim
COPY --from=build /go/bin/linux_arm64/mage /bin/mage
```
```
$ docker build . -t mage-arm64
...
Successfully tagged mage-arm64:latest
```
```
$ docker run -it --rm mage-arm64:latest /bin/mage
WARNING: The requested image's platform (linux/arm64) does not match the detected host platform (linux/amd64) and no specific platform was requested
standard_init_linux.go:219: exec user process caused: exec format error
```
```
$ docker inspect mage-arm64:latest | grep -A1 Architecture
        "Architecture": "arm64",
        "Os": "linux",
```

The advantage of this approach is that we could continue cross-building all the packages from any machine supporting a linux/amd64 docker runtime. The same way that we cross-build other packages. Also, a similar approach could also work for windows images, something that we would like to support sooner than later (https://github.com/elastic/beats/issues/16814).

The disadvantage is that we would need to refactor the Dockerfiles to prepare all files in previous stages, and to avoid running any command with `RUN` in the last platform-specific stage.

**Comment by kvch** [comment]
The current approach gets the job done, and users can have their ARM-based images. The proposal described by @jsoriano is a good enhancement for the future. I have opened a follow-up issue, so we won't forget about it: https://github.com/elastic/beats/issues/23775

We agreed offline, that this PR is a good solution for now. In the future, there is room for improvement, but it requires the refactoring of the whole Docker image related codebase. Ao we are sticking with this.

File: .ci/packaging.groovy
Comment (mdelapenya) [inline .ci/packaging.groovy:180]: Mmm we are not supporting the E2E tests for ARM yet. Once this PR is merged and we are pushing ARM-based images to our registries, then we can enable them again. We have to do some preparation in the E2E side, like selecting the target platform based on how they are called. If you all agree, we can remove this line, and we will contribute it back once ready.
```suggestion
```

File: .ci/packaging.groovy
Comment (v1v) [inline .ci/packaging.groovy:159]: Since docker login is required in ARM then we need to use a specific worker for the time being:

```suggestion
                agent { label 'worker-0eaeecf32aca4fe3a' } // TODO: use 'arm' once the problem has been fixed in all the ARM workers.
```

For instance, golang-crossbuild uses the above label, to be changed in the future (see - https://github.com/elastic/golang-crossbuild/pull/77)

File: dev-tools/mage/crossbuild.go
Comment (jsoriano) [inline dev-tools/mage/crossbuild.go:192]: Out of curiosity, can `runtime.GOARCH` be empty?

File: dev-tools/mage/crossbuild.go
Comment (jsoriano) [inline dev-tools/mage/crossbuild.go:210]: Is it possible to publish docker images for different architectures with the same tag (with multi-architecture manifests), if we do this we don't need to add logic here to chose the image depending on the local architecture, docker does it.

We can go on as this by now, but I'd suggest to publish multi-architecture manifests for maintainability.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:87]: Architecture dependant `jq` is included some lines above, we should do something similar. `jq` is used by the `docker-entrypoint` script of this image.

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:101]: Node.js is installed for heartbeat images some lines below, if we want to support synthetics in ARM we would also need to adapt this code to use ARM binaries. If we don't want to support synthetics we should add the condition to don't try to install node as is done for the `ubi-minimal` images.

**Review by jsoriano — CHANGES_REQUESTED** [review]
Elastic Agent image installs and uses `jq`, we should install it for ARM.

Same thing for nodejs in Heartbeat images, if we want to support synthetics in ARM images.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (jsoriano) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:87]: As discussed offline, this is only done in `ubi-minimal` images, so it may not be needed.

Comment (kvch) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:87]: It is installed using `yum` which loads the appropriate version of `jq`.

File: dev-tools/mage/crossbuild.go
Comment (kvch) [reply dev-tools/mage/crossbuild.go:192]: No, I've just had a different code here, and did not adjust the check after the refactoring.

File: .ci/packaging.groovy
Comment (kvch) [reply .ci/packaging.groovy:175]: Rename packaging notification

File: .ci/packaging.groovy
Comment (v1v) [reply .ci/packaging.groovy:169]: This is not required anymore https://github.com/elastic/apm-pipeline-library/pull/928 is now in charge to set the arch for installing go and so on in case GOARCH was not set.

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:154]: Use the same version for both archs, there is already a `NODE_VERSION` variable that could be used:

```suggestion
       x86_64) \
           NODE_DOWNLOAD_URL=https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-x64.tar.xz \
           ;; \
       aarch64) \
           NODE_DOWNLOAD_URL=https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-arm64.tar.xz \
```

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:148]: `NODE_VERSION` is already defined some lines above, do we need to redefine it?

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:149]: Nit. Not needed.
```suggestion
```

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (jsoriano) [inline dev-tools/packaging/templates/docker/Dockerfile.tmpl:39]: I wonder why we need to specify the architecture in these packages for amd64 but not for arm64. Do we actually need this? If not this would be simplified.

**Review by v1v — CHANGES_REQUESTED** [review]
`mage package` generates build artifacts as root therefore it won't work with static workers, I'm afraid ARM workers are static yet.

For instance, the CI builds failed with the below error:
```
09:24:18  jenkins.util.io.CompositeIOException: Unable to remove directory /var/lib/jenkins/workspace/Beats_packaging_PR-23592/src/github.com/elastic/beats/auditbeat/build with directory contents: [/var/lib/jenkins/workspace/Beats_packaging_PR-23592/src/github.com/elastic/beats/auditbeat/build/golang-crossbuild]
```

So I ssh onto one of those ARM workers and `golang-crossbuild` folder is owned by `root`:

```
jenkins@worker-0eaeecf32aca4fe3a:~/workspace/Beats_packaging_PR-23592/src/github.com/elastic/beats/auditbeat/build$ ls -ltra
total 12
drwxr-xr-x 2 root    root    4096 Feb  2 09:16 golang-crossbuild
drwxr-xr-x 3 jenkins jenkins 4096 Feb  2 09:16 .
```

This PR doesn't work with the existing CI infrastructure unless the permissions are changed

**Comment by kvch** [comment]
UBI support needs to be resolved.

File: dev-tools/mage/dockerbuilder.go
Comment (michalpristas) [inline dev-tools/mage/dockerbuilder.go:75]: this is funny, i like it

**Review by michalpristas — APPROVED** [review]
from my point of view it looks ok, maybe it would be ok to have additional approval from somebody involved in infra/build for longer period of time

**Comment by mdelapenya** [comment]
The failed E2E tests seems related to a code base not rebased/merged with master branch: there was a fix yesterday related to them

File: dev-tools/packaging/templates/docker/Dockerfile.tmpl
Comment (andrewvc) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl:39]: I think we can  remove this. These are all dependencies for  chromium, which is used for synthetics. While synthetics probably won't work out of the box on arm till figure out how to get arm chromium builds there's no reason not to remove the arch from these packages.

Comment (andrewvc) [reply dev-tools/packaging/templates/docker/Dockerfile.tmpl:39]: By remove, btw, I don't mean the packages, but the arch suffix

File: x-pack/elastic-agent/magefile.go
Comment (jsoriano) [inline x-pack/elastic-agent/magefile.go:595]: Should we do this in any case if docker is one of the selected packages? also if there are more than one selected package.

File: dev-tools/packaging/packages.yml
Comment (jsoriano) [inline dev-tools/packaging/packages.yml:934]: Is `jq` for ARM in the ubi image needed at the end? we are installing the "linux64" version now.

File: x-pack/elastic-agent/magefile.go
Comment (kvch) [reply x-pack/elastic-agent/magefile.go]: ATM it's not necessary. But for the sake of being future proof, I have rewritten the code.

File: dev-tools/packaging/packages.yml
Comment (kvch) [reply dev-tools/packaging/packages.yml:934]: Now we are installing the appropriate version.

**Review by v1v — APPROVED** [review]

**Comment by kvch** [comment]
Packaging has been successful on multiple ocasions.

**Comment by kvch** [comment]
jenkins run tests

**Comment by kvch** [comment]
Tests have passed.

**Pull request merged**
